technote about `@owner` patterns (#376)
apollographql/docs: src/content/technotes/TN0030-owner-pattern.mdx

---

> title: Ownership in Federation

Reviewer: ```suggestion
title: Enforcing entity ownership in Apollo Federation
```

```
HINT: [INCONSISTENT_DESCRIPTION]: Element "Product" has inconsistent
descriptions across subgraphs. The supergraph will use description
(from subgraph "one"):
  """
  The Product type lorem ipsum dolar sit amet.
  """
In subgraph "two", the description is:
  """
  This is my description of the Product type.
  """
```

> The "winner" is the subgraph that comes first alphabetically by name.

A mechanism for deciding the "owner" of the type allows tools such as linters to catch these inconsistencies early in the development process.

## Creating an `@owner` directive

You can add an `@owner` directive to your supergraph using the [`@composeDirective`](/federation/federated-types/federated-directives#composedirective) functionality introduced in Federation 2.2.

```graphql title="subgraph-a.graphql"
extend schema
  @link(
    url: "https://specs.apollo.dev/federation/v2.3"
    import: ["@key", "@composeDirective"]
  )
  @link(url: "https://graphql.mycompany.dev/owner/v1.0")
  @composeDirective(name: "@owner")

directive @owner(team: String!) on OBJECT

type Product @key(fields: "id") @owner(team: "subgraph-a") {
  id: ID!
  name: String
}
```

> The `@owner` directive now appears in the supergraph. Because we did not define the directive as `repeatable`, subgraphs cannot define it with different arguments.

Author: Almost! Multiple subgraphs can define the same directive as long as the arguments are the same.

directive @join__implements(
  graph: join__Graph!
  interface: String!
) repeatable on OBJECT | INTERFACE

directive @join__type(
  graph: join__Graph!
  key: join__FieldSet
  extension: Boolean! = false
  resolvable: Boolean! = true
  isInterfaceObject: Boolean! = false
) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR

directive @join__unionMember(
  graph: join__Graph!
  member: String!
) repeatable on UNION

directive @link(
  url: String
  as: String
  for: link__Purpose
  import: [link__Import]
) repeatable on SCHEMA

directive @owner(team: String!) on OBJECT

scalar join__FieldSet

enum join__Graph {
  ONE @join__graph(name: "one", url: "http://localhost:4001")
}

scalar link__Import

enum link__Purpose {
  """
  `SECURITY` features provide metadata necessary to securely resolve fields.
  """
  SECURITY

  """
  `EXECUTION` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

type Product @join__type(graph: ONE, key: "id") @owner(team: "subgraph-a") {
  id: ID!
  name: String
}

type Query @join__type(graph: ONE) {
  products: [Product]
}
```

</ExpansionPanel>

## Writing a lint rule using the `@owner` directive

> Here is an example of a [`@graphql-eslint`](https://the-guild.dev/graphql/eslint/docs) rule for subgraph schemas that uses the `@owner` directive to determine if a description is required.

Reviewer: ```suggestion
Here's an example of a [`@graphql-eslint`](https://the-guild.dev/graphql/eslint/docs) rule for subgraph schemas that uses the `@owner` directive to determine if a description is required:
```

> <ExpansionPanel title="Click to view rule code">

Reviewer: ```suggestion
<ExpansionPanel title="Click to expand">
```

        const schema = buildSchema(context.getSourceCode().text, {
          assumeValidSDL: true // subgraph schemas may not be valid on their own
        });

        return {
          /**
           * For each object type defintion, look for an `@owner` directive.
           * If it exist, require a description.
           * If it doesn't, disallow a description.
           */
          ObjectTypeDefinition(node) {
            const type = schema.getType(node.name.value);
            const owner = getDirective(schema, type, 'owner');

            if (owner && !node.description) {
              context.report({
                node,
                message: 'Description is required on owned types'
              });
            }

            if (!owner && node.description) {
              context.report({
                node,
                message: 'Description not allowed on unowned types'
              });
            }
          }
        };
      }
    }
  }
};
```

</ExpansionPanel>

## Using `@owner` to determine required approvers

Another use case for the `@owner` directive is to determine required reviewers when a schema change affects a type owned by another team.

> The exact process will depend on your source control and continuous integration systems. The following example steps assume you're using GitHub for both.

Reviewer: ```suggestion
The exact process depends on your source control and continuous integration systems. The following example steps assume you're using GitHub for both.
```

1. Add a `pull_request` workflow:

> ```yaml title=".github/workflows/add-reviewers.yaml"
> name: Add required reviewers for owned GraphQL types
> on: [pull_request]
> ```

Reviewer: ````suggestion
    ```yaml title=".github/workflows/add-reviewers.yaml"
    name: Add required reviewers for owned GraphQL types
    on: [pull_request]
    ```
````

2. Determine the affected types in the schema change:

> ```js

Reviewer: Also indent this block 4 spaces

import {diff} from '@graphql-inspector/core';
import {buildSchema} from 'graphql';

const differences = diff(
  buildSchema(schemaFromBaseRef, {assumeValidSDL: false}),
  buildSchema(schemaFromCurrentRef, {assumeValidSDL: false})
);

/* result:
[
  {
    "criticality": {
      "level": "NON_BREAKING"
    },
    "type": "FIELD_ADDED",
    "message": "Field 'newField' was added to object type 'Product'",
    "path": "Product.newField"
  }
]
*/
```

> 3. Obtain the supergraph schema:

Reviewer: ```suggestion
3. Obtain the supergraph schema.
```

> You can use [`rover supergraph fetch`](https://www.apollographql.com/docs/rover/commands/supergraphs#supergraph-fetch) or retreive it using the [Apollo Platform API](https://studio.apollographql.com/public/apollo-platform/explorer?explorerURLState=N4IgJg9gxgrgtgUwHYBcQC4QEcYIE4CeABAOIIoDKMADvgOZ4CG1AFgBQAkeCAZukQEkAIgEIAlEWAAdJESIA3RngCWjVG258iXXhOmy5RAPpGUBWkkaIZhogDoHRCLJJNWANSWrUkm7aIANowoCADOKACC1NR4EPIIYAAyjDBIUCy%2BBv5EAEYwygFgmdm23KEwASjFJbYmZhZWCH41cg52TrIAQvmFVFBQYaHVLYZQENwU6QhwjMMjo%2BMIQtDwyCjN8wC%2BGzXbWSV7uztEh4ane5sgm0A&variant=main).

Reviewer: Indent 4 spaces

Reviewer: (same for remaining post-list-item codeblocks below)
